Test/ajodwyer/rcu_cell.hpp

```diff
@@ -0,0 +1,231 @@
+        using cb_allocator = typename std::allocator_traits<Alloc>::template rebind_alloc<cell_control_block>;
+        T *t;
+        cb_allocator a;
+        explicit cell_control_block(T *t, Alloc a) : t(t), a(std::move(a)) {}
+    };
+} // namespace detail
+
+template <typename T, typename Alloc = std::allocator<T>>
+class cell {
+    using control_block = detail::cell_control_block<T, Alloc>;
+    using cb_allocator = typename std::allocator_traits<Alloc>::template rebind_alloc<control_block>;
+    using cb_pointer = typename std::allocator_traits<cb_allocator>::pointer;
+    std::atomic<control_block *> cb;
+    cb_allocator a;
+
+    static_assert(std::is_same<
+        typename std::allocator_traits<cb_allocator>::pointer,
+        control_block *
+    >::value, "cell<T,A> requires that A::rebind_alloc<CB>::pointer be exactly CB*");
+
+  public:
+    cell(cell&&) = delete;
+    cell(const cell&) = delete;
+    cell& operator=(cell&&) = delete;
+    cell& operator=(const cell&) = delete;
+
+    cell() : cb(nullptr), a(Alloc()) {}
+    explicit cell(nullptr_t, Alloc alloc = Alloc()) : cb(nullptr), a(std::move(alloc)) {}
+    explicit cell(std::unique_ptr<T> u, Alloc alloc = Alloc()) : a(std::move(alloc)) {
+        control_block *new_cb = std::allocator_traits<cb_allocator>::allocate(a, 1);
+        std::allocator_traits<cb_allocator>::construct(a, new_cb, u.release(), a);
+        cb = new_cb;
+    }
+    ~cell() {
+        control_block *old_cb = cb;
+        if (old_cb) old_cb->retire();
+    }
+
+    void update(nullptr_t) {
+        // "Update" the cell to become "empty", which means to abandon (retire) it.
+        control_block *old_cb = cb.exchange(nullptr);
+        if (old_cb) old_cb->retire();
+    }
+
+    void update(unique_ptr<T> u) {
+        // "Update" the cell to contain a new value, which means to allocate
+        // a new cell and abandon (retire) the old one.
+        if (u == nullptr) {
+            this->update(nullptr);
+        } else {
+            control_block *new_cb = std::allocator_traits<cb_allocator>::allocate(a, 1);
+            std::allocator_traits<cb_allocator>::construct(a, new_cb, u.release(), a);
+            control_block *old_cb = cb.exchange(new_cb);
+            if (old_cb) old_cb->retire();
+        }
+    }
+
+    snapshot_ptr<T> get_snapshot() const {
+        ::rcu_read_lock();
+        control_block *old_cb = cb;
+        if (old_cb) {
+            return snapshot_ptr<T>(old_cb->t);
+        } else {
+            ::rcu_read_unlock();
+            return nullptr;
+        }
+    }
+};
+
+template <typename T>
+class snapshot_ptr {
+    friend class cell<T>;
+    T *ptr;
+
+    explicit snapshot_ptr(T *p) : ptr(p) {
+        assert(p != nullptr);
+    }
+
+  public:
+    constexpr snapshot_ptr(snapshot_ptr&& rhs) : ptr(rhs.ptr) {
+        rhs.ptr = nullptr;
+    }
+    snapshot_ptr& operator=(snapshot_ptr&& rhs) {
+        if (ptr) {
+            ::rcu_read_unlock();
```

Comment:
I think there's a high-level design problem here. Suppose I have this:

    cell<A> c(make_unique<A>(1));
    auto sp1 = c.get_snapshot(); // calls ::rcu_read_lock()
    c.update(make_unique<A>(2));
    ::sleep(1);
    auto sp2 = c.get_snapshot(); // calls ::rcu_read_lock()
    ::sleep(1);
    sp1 = std::move(sp2);  // calls ::rcu_read_unlock()
    ::sleep(1000);

During the sleep(1000), we humans know that the object `A(1)` is unreferenced, thus cleanuppable; and we know that the object `A(2)` is still referenced by `sp1`. But all the machine knows is that the count of outstanding `rcu_read_lock` calls in this thread is still 1 (down from 2 a moment earlier). So the machine isn't going to try to clean up *anything*, right?

So holding one `snapshot_ptr` in a global singleton or something might end up disabling RCU garbage collection forever?

I think this is related to the "cookie" idea.